pkg/util/util.go

```
@@ -239,3 +244,147 @@ func SplitQualifiedName(str string) (string, string) {
+				return nil, err
+			}
+			//Only IPv4
+			//TODO : add IPv6 support
+			if ip.To4() != nil {
+				if !ip.IsLoopback() {
+					glog.V(4).Infof("IP found %v\n", ip)
+					return ip, nil
+				} else {
+					glog.V(4).Infof("Loopback found %v\n", ip)
+				}
+			} else {
+				glog.V(4).Infof("%v is not a valid IPv4 address\n", ip)
+			}
+
+		}
+	}
+	return nil, nil
+}
+
+func getIPFromInterface(intfName string) (net.IP, bool, error) {
+	intf, err := net.InterfaceByName(intfName)
+	if err != nil {
+		return nil, false, err
+	}
+	if isInterfaceUp(intf) {
+		addrs, err := intf.Addrs()
+		if err != nil {
+			return nil, false, err
+		}
+		glog.V(4).Infof("Interface %q has %d addresses :%v.\n", intfName, len(addrs), addrs)
+		finalIP, err := getFinalIP(addrs)
+		if err != nil {
+			return nil, false, err
+		}
+		if finalIP != nil {
+			glog.V(4).Infof("valid IPv4 address for interface %q found as %v.\n", intfName, finalIP)
+			return finalIP, true, nil
+		}
+	}
+
+	return nil, false, nil
+}
+
+//ChooseHostInterface is a method used fetch an IP for a daemon.
+//It uses data from /proc/net/route file.
+//For a node with no internet connection ,it returns 127.0.0.1 and the user has to set the IP to 127.0.0.1.
+//For a multi n/w interface node it returns the IP of the interface with gateway on it.
+func ChooseHostInterface() (net.IP, bool, error) {
+	zero := net.IP{0, 0, 0, 0}
+	localhost := net.IP{127, 0, 0, 1}
+	routes, err := getRoutes("/proc/net/route")
+	if err != nil {
+		return nil, false, err
+	}
+	for i := range routes {
+		//find interface with gateway
+		if routes[i].Destination.Equal(zero) {
+			glog.V(4).Infof("Default route transits interface %q\n", routes[i].Interface)
+			ip, found, err := getIPFromInterface(routes[i].Interface)
+			if err != nil {
+				return nil, false, err
+			}
+			if found {
+				glog.V(4).Infof("Choosing IP %v \n", ip)
+				return ip, true, nil
+			}
+		}
+	}
+	glog.V(4).Infof("No valid IP found\n")
+	return localhost, false, nil
```

I think for now we should return an error, which means no need for localhost variable